varname/ignore.py

```diff
@@ -0,0 +1,343 @@
+   and recommended when code is executed by `exec` without module available.
+2. Ignore frames by a given pair of module and a qualified name (qualname).
+   See 1) for acceptable modules. The qualname should be unique in that module.
+3. Ignore frames by a (non-decorated) function.
+4. Ignore frames by a decorated function. In this case, you can specified a
+   tuple with the function and the number of decorators of it. The decorators
+   on the wrapper function inside the decorators should also be counted.
+
+Any frames in `varname` , and frames of any expressions like
+<listcomp/dictcomp/setcomp/genexpr/lambda> are ignored by default.
+
+Attributes:
+
+    MODULE_IGNORE_ID_NAME: The name of the ignore id injected to the module.
+        Espectially for modules that can't be retrieved by
+        `inspect.getmodule(frame)`
+"""
+import sys
+import re
+import inspect
+import distutils.sysconfig as sysconfig
+import warnings
+from os import path
+from abc import ABC, abstractmethod
+from typing import List, Union, Tuple, Optional
+from types import FrameType, ModuleType, FunctionType
+
+from executing import Source
+
+IgnoreElemType = Union[
+    # module
+    ModuleType,
+    # filename of a module
+    str,
+    FunctionType,
+    # the module (filename) and qualname
+    # If module is None, then all qualname matches the 2nd element
+    # will be ignored. Used to ignore <listcomp/dictcomp/setcomp/genexpr>
+    # internally
+    Tuple[Optional[Union[ModuleType, str]], str],
+    # Function and number of its decorators
+    Tuple[FunctionType, int]
+]
+IgnoreType = Union[IgnoreElemType, List[IgnoreElemType]]
+
+MODULE_IGNORE_ID_NAME = '__varname_ignore_id__'
+
+def debug_ignore_frame(msg: str,
+                       frameinfo: Optional[inspect.FrameInfo] = None) -> None:
+    """Print the debug message for a given frame info object
+
+    Args:
+        msg: The debugging message
+        frameinfo: The FrameInfo object for the frame
+    """
+    from .utils import config
+    if not config.debug:
+        return
+    if frameinfo is not None:
+        msg = (f'{msg} [In {frameinfo.function!r} at '
+               f'{frameinfo.filename}:{frameinfo.lineno}]')
+    sys.stderr.write(f'[{__name__}] DEBUG: {msg}\n')
+
+class IgnoreElem(ABC):
+    """An element of the ignore list"""
+
+    def __init__(self, ignore: IgnoreElemType) -> None:
+        self.ignore = ignore
+
+    @abstractmethod
+    def match(self, frame_no: int, frameinfos: List[inspect.FrameInfo]) -> bool:
+        """Whether the frame matches the ignore element"""
+
+    @abstractmethod
+    def __repr__(self) -> str:
+        """Representation of the element"""
+
+class IgnoreModule(IgnoreElem):
+    """Ignore calls from a module or its submodules"""
+
+    def __init__(self, ignore: Union[ModuleType, str]) -> None:
+        super().__init__(ignore)
+        self.is_module = isinstance(ignore, ModuleType)
```

Comment:
Seems like you should just have two subclasses

Reply:
Not quite clear about what you are saying. Subclasses of `IgnoreElem`?

Reply:
Yes

Reply:
So one dealing with module, the other function?

Reply:
One for module, one for filename. Instead of constant `if self.is_module`.

Reply:
Oh, you mean to split this `IgnoreModule` into `IgnoreModule` and `IgnoreFilename`?